File: 프로그래머스/LV3/코딩테스트 연습/2020 카카오 인턴십/경주로 건설.py
from collections import deque

def solution(board):
    N = len(board)
    loc_cost = [[[25 * 25 * 500 for _ in range(4)] for _ in range(N)] for _ in range(N)]
    ds = [[1,0], [0,1], [-1,0], [0,-1]]
    q = deque([[0, 0, 0, 0], [0, 0, 1, 0]])

    while q:
        r, c, pdi, cost = q.popleft()

        for di, d in enumerate(ds):
            next_r, next_c = r + d[0], c + d[1]
            if next_r < 0 or next_r >= N or next_c < 0 or next_c >= N: continue
            if board[next_r][next_c] == 1: continue

            new_cost = cost + 100 if di == pdi else cost + 600

            if new_cost < loc_cost[next_r][next_c][di]:
                loc_cost[next_r][next_c][di] = new_cost
                q.append([next_r, next_c, di, new_cost])

    return min(loc_cost[N-1][N-1])

File: 프로그래머스/LV3/코딩테스트 연습/2020 카카오 인턴십/경주로 건설.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

class P {
public:
    int r = 0;
    int c = 0;
    int pdi = 0;
    int cost = 0;

    P(int _r, int _c, int _pdi, int _cost) {
        this->r = _r;
        this->c = _c;
        this->pdi = _pdi;
        this->cost = _cost;
    }
};

int solution(vector<vector<int>> board) {
    int N = board.size();
    vector<vector<vector<int>>> locCost;
    for (int r = 0; r < N; r++) {
        vector<vector<int>> row;
        for (int c = 0; c < N; c++) {
            vector<int> temp(4, 25 * 25 * 500);
            row.push_back(temp);
        }
        locCost.push_back(row);
    }

    vector<vector<int>> ds = { {1, 0}, {0, 1}, {-1, 0}, {0, -1} };
    queue<P> q;
    q.push(P(0, 0, 0, 0));
    q.push(P(0, 0, 1, 0));

    while (!q.empty()) {
        P p = q.front();
        q.pop();

        for (int di = 0; di < 4; di++) {
            int nextR = p.r + ds[di][0];
            int nextC = p.c + ds[di][1];
            if (nextR < 0 || nextR >= N || nextC < 0 || nextC >= N) continue;
            if (board[nextR][nextC] == 1) continue;

            int newCost = di == p.pdi ? p.cost + 100 : p.cost + 600;

            if (newCost < locCost[nextR][nextC][di]) {
                locCost[nextR][nextC][di] = newCost;
                q.push(P(nextR, nextC, di, newCost));
            }
        }
    }
    return *min_element(locCost[N-1][N-1].begin(), locCost[N - 1][N - 1].end());;
}